src/Build.UnitTests/WarningsAsMessagesAndErrors_Tests.cs

```diff
@@ -272,6 +272,34 @@ private string GetTestProject(bool? treatAllWarningsAsErrors = null, string warn
             </Project>";
         }
 
+        [Fact]
+        public void WarningsAsErrors_ExpectBuildToStopWhenTaskLogsWarningAsError()
+        {
+            using (TestEnvironment env = TestEnvironment.Create(_output))
+            {
+                TransientTestProjectWithFiles proj = env.CreateTestProjectWithFiles($@"
+                <Project>
+                    <UsingTask TaskName = ""ReturnFailureWithoutLoggingErrorTask"" AssemblyName=""Microsoft.Build.Engine.UnitTests""/>
+                    <UsingTask TaskName = ""LogWarningReturnHasLoggedError"" AssemblyName=""Microsoft.Build.Engine.UnitTests""/>
+                    <PropertyGroup>
+                        <MSBuildWarningsAsErrors>MSB1234</MSBuildWarningsAsErrors>
+                    </PropertyGroup>
+                    <Target Name='Build'>
+                        <LogWarningReturnHasLoggedError WarningCode=""MSB1234""/>
+                        <ReturnFailureWithoutLoggingErrorTask/>
+                    </Target>
+                </Project>");
+
+                MockLogger logger = proj.BuildProjectExpectFailure();
+
+                logger.WarningCount.ShouldBe(0);
+                logger.ErrorCount.ShouldBe(1);
+
+                // The build should STOP when a task logs an error, make sure ReturnFailureWithoutLoggingErrorTask doesn't run. 
+                logger.AssertLogDoesntContain("MSB4181");
```

Comment:
I think this may not work with tasks that implement ITask directly. Can you please add another test just like this one but calling a task implementing ITask directly?

Reply:
Interesting. Going to brain dump this before making the test.

Someone implementing ITask wouldn't have access to a TaskLoggingHelper (like Task does) and thus have to log a warning through the `BuildEngine.LogWarningEvent`. The build engine would log it to the logging service as a warning and at `LoggingService.RouteBuildEvent` it would be translated into an error.

But I think the build would continue because it entirely depends on what the task returned right? So long as the task returns true despite logging an error, the build will complete.

Question, what context would someone implement itask vs inheriting from task?

Reply:
Don't really know why people would do it, but according to this [GH search](https://sourcegraph.com/search?q=lang%3Acsharp+%28content%3A%22%3A+ITask+%22+or+content%3A%22%3A+ITask%5Cn%22%29+and+content%3A%22using+Microsoft.Build.Framework%22+count%3A1000&patternType=literal) they do it.

> So long as the task returns true despite logging an error, the build will complete.

I think MSBuild might either warn or error that the task returned true but an error was logged. Not super sure about this though, so worth writing a test to pin the behaviour :)
On the other hand, since an ITask implementation cannot know whether any errors were logged (because it does not have access to `Log`), maybe this whole scenario does not make sense and we can ignore it.

Reply:
I agree that there's no need to check this on an ITask. Something interesting about the error task. It explicitly returns false otherwise the build would continue: https://github.com/dotnet/msbuild/blob/f0eebf2872d76ab0cd43fdc4153ba636232b222f/src/Tasks/Error.cs#L47 that's a very strange disconnect that sounds to me like it should be an issue to fix. But if that's just how msbuild has been since the beginning of time...should we consider fixing it?

Reply:
It comes from an era where an error was an error, so returning false makes sense :)

I'd wait until somebody reports an issue and then update both Error and Warning.

Reply:
Sorry, why is that wrong? A task returns false if it logged an error. The Error task is designed to log an error, hence always returns "HasLoggedAnError."

Reply:
I should have included it in the code sample, but the error task called `Log.LogError` and simply returns false, not `Log.HasLoggedErrors`.

Reply:
Right...it has always logged an error because it logs an error right before the return statement.

Right, and people who log errors tend to want execution to stop at that point. No point continuing if something went irrevocably wrong.